music21/musicxml/m21ToXml.py

```diff
@@ -1672,6 +1743,186 @@ def setScoreHeader(self):
         # the hard one...
         self.setPartList()
 
+    def joinPartStaffs(self):
+        '''
+        Collect <part> elements exported from
+        :class:`~music21.stream.PartStaff` objects under a single
+        <part> element using <staff> and <voice> subelements.
+
+        >>> from music21.musicxml import testPrimitive
+        >>> s = converter.parse(testPrimitive.pianoStaff43a)
+        >>> SX = musicxml.m21ToXml.ScoreExporter(s)
+        >>> root = SX.parse()
+        >>> parts = root.findall('part')
+        >>> len(parts)
+        1
+        >>> staffTags = root.findall('part/measure/note/staff')
+        >>> len(staffTags)
+        2
+        '''
+        DIVIDER_COMMENT = '========================= Measure n =========================='
+
+        staffGroups = self.stream.getElementsByClass('StaffGroup')
+        joinableGroups = []
+        for sg in staffGroups:
+            # Joinable groups must consist of only PartStaffs with Measures
+            if (
+                len(sg) > 1
+                and all(isinstance(p, stream.PartStaff) for p in sg)
+                and all(p.getElementsByClass('Measure') for p in sg)
+            ):
+                joinableGroups.append(sg)
+
+        # Pass 1: create <staff> tags and move elements
+        for group in joinableGroups:
+            initialRoot = None
+            for i, ps in enumerate(group):
+                staffNumber = i + 1  # 1-indexed
+                root = self._getRootForPartStaff(ps)
+
+                # Create <staff> tags under <note>, <direction>, <forward> tags
+                for mxMeasure in root.findall('measure'):
+                    XMLExporterBase.addStaffTag(mxMeasure, staffNumber,
+                        tagList=['note', 'direction', 'forward'])
+
+                # Move elements
+                if initialRoot is None:
+                    initialRoot = root
+                else:
+                    initialHigh = max(int(m.get('number')) for m in initialRoot.findall('measure'))
+                    thisHigh = max(int(m.get('number')) for m in root.findall('measure'))
+                    highestMeasureNumber = max(initialHigh, thisHigh)
+                    initialRootCursor = 0
+                    for mNum in range(highestMeasureNumber + 1):
+                        initialMeasure = initialRoot.find(f"measure[@number='{mNum}']")
+                        thisMeasure = root.find(f"measure[@number='{mNum}']")
+                        if thisMeasure is None and initialMeasure is None:
+                            # Gap in both measure sequences
+                            continue
+                        if thisMeasure is None:
+                            # Gap in this measure sequence
+                            initialRootCursor += 2  # Advance for comment & measure in initial seq.
```

Comment:
I probably need to refactor this to use three iterators, for element in initial, for element in this, for i in range(highestMeasureNum + 1), since I probably shouldn't depend on measure numbers being strictly numeric (suffixes, etc.)

Reply:
Gritted my teeth and decided to walk through the <measure> tags in the target <part> and the music21 elements in the PartStaffs being merged while also accounting for measure suffixes. This ballooned the complexity but is hopefully going to catch more cases. Needs some more code comments but wanted to push it up so you could see where this was headed. LMK how it looks; I'm all ears for suggestions to simplify it.